File: client/lib/services/presence_service.dart
import 'dart:async';
import '../network/socket_client.dart';
import '../network/api_client.dart';

class PresenceService {
  final SocketClient socketClient;
  final ApiClient apiClient;
  
  StreamSubscription? _presenceSub;
  
  // Emits the userId and their presence data
  final _presenceStateController = StreamController<Map<String, dynamic>>.broadcast();
  Stream<Map<String, dynamic>> get presenceState => _presenceStateController.stream;

  // In-memory cache of presence
  final Map<String, Map<String, dynamic>> _presenceCache = {};

  PresenceService({
    required this.socketClient,
    required this.apiClient,
  }) {
    _presenceSub = socketClient.presenceEvents.listen(_handleIncomingPresence);
  }

  void _handleIncomingPresence(Map<String, dynamic> payload) {
    // payload structure from Phoenix Presence diff
    final joins = payload['joins'] as Map<String, dynamic>?;
    final leaves = payload['leaves'] as Map<String, dynamic>?;

    if (joins != null) {
      for (var userId in joins.keys) {
        final state = {
          'userId': userId,
          'online': true,
          'lastSeenAt': null, // Online right now
        };
        _presenceCache[userId] = state;
        _presenceStateController.add(state);
      }
    }

    if (leaves != null) {
      for (var userId in leaves.keys) {
        // Technically they might still be joined on another device, but for v1 we'll treat a leave as offline
        // and fetch fresh status from API to be sure.
        // We add a short delay to allow the server to write the disconnect timestamp to the DB
        Future.delayed(const Duration(milliseconds: 1000), () {
          _checkStatusFallback(userId);
        });
      }
    }
  }

  Future<void> _checkStatusFallback(String userId) async {
    try {
      final data = await apiClient.fetchUserStatus(userId);
      
      final state = {
        'userId': userId,
        'online': data['online'] == true,
        'lastSeenAt': data['last_seen_at'], // ISO8601 string or null
      };
      
      _presenceCache[userId] = state;
      _presenceStateController.add(state);
    } catch (e) {
      // Ignore API errors, default to offline
      final state = {
        'userId': userId,
        'online': false,
        'lastSeenAt': null,
      };
      _presenceCache[userId] = state;
      _presenceStateController.add(state);
    }
  }

  Future<Map<String, dynamic>> getPresence(String userId) async {
    if (_presenceCache.containsKey(userId)) {
      // Return cached state but kick off an async refresh if offline
      if (_presenceCache[userId]!['online'] == false) {
        _checkStatusFallback(userId);
      }
      return _presenceCache[userId]!;
    }

    // Await first fetch
    await _checkStatusFallback(userId);
    return _presenceCache[userId]!;
  }

  void dispose() {
    _presenceSub?.cancel();
    _presenceStateController.close();
  }
}

File: android_client/app/src/main/java/com/subhankar/aurachat/service/PresenceService.kt
package com.subhankar.aurachat.service

import android.util.Log
import com.subhankar.aurachat.network.ApiClient
import com.subhankar.aurachat.network.PhoenixSocketManager
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Presence Service — replaces Dart presence_service.dart.
 *
 * Tracks online/offline status of contacts via:
 *   1. Real-time Phoenix Presence diffs (joins/leaves)
 *   2. REST API fallback for last_seen_at timestamp
 *   3. In-memory cache for instant UI reads
 */
@Singleton
class PresenceService @Inject constructor(
    private val socketManager: PhoenixSocketManager,
    private val apiClient: ApiClient
) {
    companion object {
        private const val TAG = "PresenceService"
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var collectJob: Job? = null

    private val _presenceState = MutableSharedFlow<PresenceEvent>(extraBufferCapacity = 64)
    val presenceState: SharedFlow<PresenceEvent> = _presenceState

    // In-memory cache
    private val presenceCache = mutableMapOf<String, PresenceEvent>()

    fun start() {
        collectJob?.cancel()
        collectJob = scope.launch {
            socketManager.presenceEvents.collect { payload ->
                val joins = payload["joins"] as? Map<*, *>
                val leaves = payload["leaves"] as? Map<*, *>

                joins?.keys?.forEach { key ->
                    val userId = key as? String ?: return@forEach
                    val event = PresenceEvent(userId, online = true, lastSeenAt = null)
                    presenceCache[userId] = event
                    _presenceState.emit(event)
                }

                leaves?.keys?.forEach { key ->
                    val userId = key as? String ?: return@forEach
                    // Delay 1s to let server write disconnect timestamp (matching Dart)
                    scope.launch {
                        delay(1_000)
                        checkStatusFallback(userId)
                    }
                }
            }
        }
    }

    private suspend fun checkStatusFallback(userId: String) {
        try {
            val data = apiClient.fetchUserStatus(userId)
            val event = PresenceEvent(
                userId = userId,
                online = data["online"] == true,
                lastSeenAt = data["last_seen_at"] as? String
            )
            presenceCache[userId] = event
            _presenceState.emit(event)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to fetch status for $userId: ${e.message}")
            val event = PresenceEvent(userId, online = false, lastSeenAt = null)
            presenceCache[userId] = event
            _presenceState.emit(event)
        }
    }

    /**
     * Get cached presence, with background refresh if offline.
     * Replaces: Dart presenceService.getPresence()
     */
    suspend fun getPresence(userId: String): PresenceEvent {
        val cached = presenceCache[userId]
        if (cached != null) {
            if (!cached.online) {
                // Kick off async refresh
                scope.launch { checkStatusFallback(userId) }
            }
            return cached
        }
        // First fetch
        checkStatusFallback(userId)
        return presenceCache[userId] ?: PresenceEvent(userId, false, null)
    }

    fun stop() { collectJob?.cancel() }
    fun dispose() { stop(); scope.cancel() }
}

data class PresenceEvent(
    val userId: String,
    val online: Boolean,
    val lastSeenAt: String?
)
